### PR TITLE
[FSI] IBQN-MVQN convergence accelerator

### PR DESCRIPTION
**Pull request by rubenzorrilla**
**Description**
This adds the IBQN-MVQN convergence accelerator to the FSI application. As it is in this PR it cannot be used in the solvers (the solvers upgrade will be done in an upcoming PR).

In any case, a simple FSI problem emulator test following the line of the already existing ones have been added. As a comment aside, these tests must be refactored to use the current structure solver and FSI solvers as soon as this is fully merged. In order to do so, we can create a "fake" fluid solver that emulates the springs.

Finally, I'd like to hear your opinion on the naming convention for the subdomains update. For the sake of generality, I decided to use `UpdateLeft` and `UpdateRight`. After having finished all the developments, I've the feelings that these are way misleading for the users. Do you think that it's better to keep those or to chance to something general but more self-descriptive like `UpdateSolutionField` and `UpdateTractionField`?

### CODE REVIEW

**Comment by loumalouomega** [comment]
IBQN == Interface Block Quasi Newton?

**Comment by rubenzorrilla** [comment]
> IBQN == Interface Block Quasi Newton?

Yes!!!! 😸

File: applications/TrilinosApplication/custom_python/add_trilinos_convergence_accelerators_to_python.cpp [hunk 1]
```diff
@@ -53,6 +53,7 @@ void AddTrilinosConvergenceAcceleratorsToPython(pybind11::module &m)
     using TrilinosBaseConvergenceAcceleratorType = ConvergenceAccelerator<TrilinosSparseSpaceType, TrilinosLocalSpaceType>;
 
     py::class_<TrilinosBaseConvergenceAcceleratorType, typename TrilinosBaseConvergenceAcceleratorType::Pointer>(m, "TrilinosConvergenceAccelerator")
+        .def(py::init<>())
```
Comment (philbucher) [inline applications/TrilinosApplication/custom_python/add_trilinos_convergence_accelerators_to_python.cpp]: I would not allow this
why do you want to add it?

IMO the only CTor we have should be the one that accepts `Parameter`s

File: applications/FSIApplication/tests/FSI_problem_emulator_test.py [hunk 2]
```diff
@@ -255,10 +401,63 @@ def ComputeDirichletNeumannResidual(self):
...
+
 if __name__ == '__main__':
     test = FSIProblemEmulatorTest()
     test.setUp()
     # test.testFSIProblemEmulatorWithAitken()
     # test.testFSIProblemEmulatorWithMVQN()
-    test.testFSIProblemEmulatorWithMVQNRecursive()
+    test.testFSIProblemEmulatorWithIBNMVQN()
+    # test.testFSIProblemEmulatorWithMVQNRecursive()
     test.tearDown()
```
Comment (philbucher) [inline applications/FSIApplication/tests/FSI_problem_emulator_test.py]: you know that there is a python way to do this?
in this case use `python3 FSI_problem_emulator_test.py FSIProblemEmulatorTest.testFSIProblemEmulatorWithIBNMVQN`

see the [python docs](https://docs.python.org/3/library/unittest.html#command-line-interface)

I use it all the time, this is super super useful

File: applications/FSIApplication/tests/FSI_problem_emulator_test.py [hunk 3]
```diff
@@ -255,10 +401,63 @@ def ComputeDirichletNeumannResidual(self):
...
+        for i in range(self.GetInterfaceProblemSize()):
+            self.iteration_value_f[2*i] = self.fluid_emulator_stiffness * self.corrected_iteration_value_u[2*i]
+            self.iteration_value_f[2*i+1] = self.fluid_emulator_stiffness * self.corrected_iteration_value_u[2*i+1]
+
+    def __TransferFluidEmulatorLoad(self):
+        interface_submodelpart = self.structure_solver.main_model_part.GetSubModelPart("StructureInterface2D_Solid_interface")
+
+        i = 0
+        for node in interface_submodelpart.Nodes:
+            point_load = Vector(3)
```
Comment (philbucher) [inline applications/FSIApplication/tests/FSI_problem_emulator_test.py:436]: just saying that you could also pass a `list` instead

File: applications/FSIApplication/tests/FSI_problem_emulator_test.py [hunk 4]
```diff
@@ -178,6 +186,144 @@ def RunTestCase(self):
...
+            for i in range(0,residual_size):
+                self.iteration_value_u[i] = 0.0
+                self.iteration_value_f[i] = 0.0
+                self.corrected_iteration_value_u[i] = 0.0
+                self.corrected_iteration_value_f[i] = 0.0
+
+            step = 0
+            time = 0.0
+
+            while(time <= self.end_time):
```
Comment (philbucher) [inline applications/FSIApplication/tests/FSI_problem_emulator_test.py:268]: cannot this be further unified? => should be the same loop for al the accelerators no?

File: applications/FSIApplication/tests/test_FSIApplication.py [hunk 5]
```diff
@@ -41,6 +41,7 @@ def AssembleTestSuites():
     smallSuite.addTest(FSICouplingInterfaceTest('test_fsi_coupling_interface'))
     smallSuite.addTest(FSIProblemEmulatorTest('testFSIProblemEmulatorWithAitken'))
     smallSuite.addTest(FSIProblemEmulatorTest('testFSIProblemEmulatorWithMVQN'))
+    smallSuite.addTest(FSIProblemEmulatorTest('testFSIProblemEmulatorWithIBNMVQN'))
```
Comment (philbucher) [inline applications/FSIApplication/tests/test_FSIApplication.py]: maybe makes more sense to add them all at once, this way new test will not be forgotten to be added

File: applications/FSIApplication/custom_utilities/mvqn_convergence_accelerator.hpp [hunk 6]
```diff
@@ -217,6 +218,8 @@ class MVQNFullJacobianConvergenceAccelerator: public ConvergenceAccelerator<TSpa
     ///@name Friends
     ///@{
 
+    friend class IBQNMVQNConvergenceAccelerator<TSparseSpace, TDenseSpace>;
```
Comment (philbucher) [inline applications/FSIApplication/custom_utilities/mvqn_convergence_accelerator.hpp:221]: 🙈 

can you add a comment as to why this is necessary?

Is this why you added the [protected interfaces in the BaseClass](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/solving_strategies/convergence_accelerators/convergence_accelerator.h#L198-L211)?

File: applications/FSIApplication/custom_python/add_convergence_accelerators_to_python.cpp [hunk 7]
```diff
@@ -66,6 +67,16 @@ void AddConvergenceAcceleratorsToPython(pybind11::module &m)
         .def(py::init<const double, const double, const bool>())
     ;
 
+    // MVQN convergence accelerator
+    typedef IBQNMVQNConvergenceAccelerator<SparseSpaceType, DenseSpaceType> IBQNMVQNConvergenceAcceleratorType;
+    typedef typename IBQNMVQNConvergenceAcceleratorType::Pointer IBQNMVQNConvergenceAcceleratorPointerType;
+    py::class_<IBQNMVQNConvergenceAcceleratorType, IBQNMVQNConvergenceAcceleratorPointerType, BaseConvergenceAcceleratorType>(m, "IBQNMVQNConvergenceAccelerator")
+        .def(py::init<Parameters>())
+        .def(py::init<const double, const double>())
```
Comment (philbucher) [inline applications/FSIApplication/custom_python/add_convergence_accelerators_to_python.cpp]:
```suggestion
```

I would not add this to new classes

File: applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+    {
+    }
+
+    explicit IBQNMVQNConvergenceAccelerator(Parameters rParameters)
+    {
+        Parameters mvqn_default_parameters(R"({
+            "solver_type"            : "IBQN_MVQN",
+            "w_0"                    : 0.825,
+            "abs_cut_off_tol"        : 1e-8,
+            "interface_block_newton" : true
```
Comment (philbucher) [inline applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h]: why is this here?
This is a property of the accelerator, nothing the user can choose
or did I misunderstand sth?

File: applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+     * @param rConvAcceleratorParameters Json string encapsulating the settings
+     */
+    explicit IBQNMVQNConvergenceAccelerator(
+        const double InitialOmega = 0.825,
+        const double AbsCutOff = 1e-8)
+        : mInitialOmega(InitialOmega)
+        , mpConvergenceAcceleratorLeft(Kratos::make_unique<MVQNType>(mInitialOmega, AbsCutOff, true))
+        , mpConvergenceAcceleratorRight(Kratos::make_unique<MVQNType>(mInitialOmega, AbsCutOff, true))
+    {
+    }
```
Comment (philbucher) [inline applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h]:
```suggestion
```

File: applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h [hunk 10]
```diff
@@ -0,0 +1,455 @@
...
+    ///@}
+    ///@name Operations
+    ///@{
+
+    void Initialize() override
+    {
+        KRATOS_TRY;
+
+        mpConvergenceAcceleratorLeft->Initialize();
+        mpConvergenceAcceleratorRight->Initialize();
```
Comment (philbucher) [inline applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h:129]: ahhhhhhhhhhhhhh
It uses two acc internally
didn't get that until now

File: applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h [hunk 11]
```diff
@@ -0,0 +1,455 @@
...
+
+            // Set the residual of the update problem to be solved
+            VectorType aux_RHS = rForceOutputVector - rIterationGuess;
+            VectorType aux_left_onto_right(problem_size);
+            VectorType displacement_iteration_update = *mpUncorrectedDisplacementVector - rDisplacementInputVector;
+            TSparseSpace::Mult(*p_inv_jac_left, displacement_iteration_update, aux_left_onto_right);
+            TSparseSpace::UnaliasedAdd(aux_RHS, 1.0, aux_left_onto_right);
+
+            // Set the LHS of the update problem to be solved
+            MatrixType aux_LHS = IdentityMatrix(problem_size, problem_size);
```
Comment (philbucher) [inline applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h:278]: ok I guess there is really no way to save this guy in terms of memory :D

Maybe makes sense to store this as a member to avoid constant reallocations?

File: applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h [hunk 12]
```diff
@@ -0,0 +1,455 @@
...
+            // Update the first iteration correction flag
+            mFirstRightCorrectionPerformed = true;
+
+        } else {
+            // Get the interface problem size
+            std::size_t problem_size = mpConvergenceAcceleratorRight->GetProblemSize();
+
+            // Get both left and inverse Jacobian
+            auto p_inv_jac_right = mpConvergenceAcceleratorRight->pGetInverseJacobianApproximation();
+            auto p_inv_jac_left = mpConvergenceAcceleratorLeft->pGetInverseJacobianApproximation();
```
Comment (philbucher) [inline applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h:206]: I am thinking if this should also be in the `public` Interface of the Baseclass
this way you wouldn't need the `friend` declarations

Are you planning on adding more MVQN types for this Acc?
If no then maybe the `friend` is enough an we can take out the confusing protected interface (as it is not needed due to `friend`

**Review by philbucher — COMMENTED** [review]
> 
> 
> **Description**
> This adds the IBQN-MVQN convergence accelerator to the FSI application. As it is in this PR it cannot be used in the solvers (the solvers upgrade will be done in an upcoming PR).
> 
> In any case, a simple FSI problem emulator test following the line of the already existing ones have been added. As a comment aside, these tests must be refactored to use the current structure solver and FSI solvers as soon as this is fully merged. In order to do so, we can create a "fake" fluid solver that emulates the springs.
> 
> Finally, I'd like to hear your opinion on the naming convention for the subdomains update. For the sake of generality, I decided to use `UpdateLeft` and `UpdateRight`. After having finished all the developments, I've the feelings that these are way misleading for the users. Do you think that it's better to keep those or to chance to something general but more self-descriptive like `UpdateSolutionField` and `UpdateTractionField`?

I don't have a strong opinion. Since this one seems to be very specific to FSI problems I could live also with the second suggestion

File: applications/TrilinosApplication/custom_python/add_trilinos_convergence_accelerators_to_python.cpp (hunk 1 above)
Comment (rubenzorrilla) [reply applications/TrilinosApplication/custom_python/add_trilinos_convergence_accelerators_to_python.cpp]: The usage will manifest in the next PR, which in order to make possible the usage of the Interface Block Quasi-Newton from Python refactors the Python solvers completely. I will also take the chance to simplify a lot the parallel solvers. These, in order to be in line with the serial ones will use the `FSICouplingInterface`.

The `FSICouplingInterface` owns the convergence accelerator as well as the interface model part used for the residual minimization. This is fine in serial, however the MPI convergence accelerators require the interface model part to be passed by constructor, meaning that we have a sort of interdependency.

In order to overcome this, I decided to create a default MPI communicator in the constructor of the structure `FSICouplingInterface`. Then, once the final convergence accelerator is created, it is assigned to the `FSICouplingInterface `.

(I think it would have been better to add this change in the next PR).

File: applications/FSIApplication/tests/FSI_problem_emulator_test.py (hunk 2 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/tests/FSI_problem_emulator_test.py]: :open_mouth: I wasn't aware. Thanks for the hint.

File: applications/FSIApplication/tests/FSI_problem_emulator_test.py (hunk 3 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/tests/FSI_problem_emulator_test.py:436]: As mentioned above this test needs to be completely refactored. I'd leave these changes for the a standalone PR doing this.

File: applications/FSIApplication/tests/FSI_problem_emulator_test.py (hunk 4 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/tests/FSI_problem_emulator_test.py:268]: Nope. If they are block Newton the current non-linear loop is not valid anymore (this is why I required to refactor all the solvers). What I'll do for the test upgrade is to use the FSI solver and create a fake fluid emulation solver with the springs.

File: applications/TrilinosApplication/custom_python/add_trilinos_convergence_accelerators_to_python.cpp (hunk 1 above)
Comment (philbucher) [reply applications/TrilinosApplication/custom_python/add_trilinos_convergence_accelerators_to_python.cpp]: could you take out this change from this PR so that we can discuss it in the upcoming one?

File: applications/FSIApplication/tests/FSI_problem_emulator_test.py (hunk 4 above)
Comment (philbucher) [reply applications/FSIApplication/tests/FSI_problem_emulator_test.py:268]: sure... 🤦

File: applications/FSIApplication/custom_utilities/mvqn_convergence_accelerator.hpp (hunk 6 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/custom_utilities/mvqn_convergence_accelerator.hpp:221]: Yeah. I was expecting this coment :smile: 

Yes. My original idea was to intentionally difficult the access to the interface Jacobians to avoid these to be misused. Since the `IBQNMVQNConvergenceAccelerator` derives from the `ConvergenceAccelerator` base class, this requires to make it a `friend` of the standard `MVQN`.

In here I see two things:
1. We could derive the `IBQNMVQN` from the `MVQN`. However this would preclude the creation of an eventual intermediate `InterfaceBlockQuasiNewtonConvergenceAccelerator` class unifying the API of all the interface block Quasi-Newton convergence accelerators.
2. Maybe it is not worth it to do this and we can just make the access to the Jacobians `public`.

File: applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h (hunk 8 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h]: You understood it correctly, it is a property of the convergence accelerator. I added it as "information".

File: applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h (hunk 10 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h:129]: Yes. One takes care of the tractions while the other one does so for the displacements.

File: applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h (hunk 11 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h:278]: :smile: 

I've discussing again with @RiccardoRossi . This is a very first working but not optimal implementation,  we wonder about the possibility of implementing the projection of the Jacobian onto a vector rather than exposing it.

In any case we think that we should first merge this (together with the changes in the solvers to enable using it) to have a framework to put the hands on. The nice thing would be that all the future improvements will be located at the "algebra" level rather than in the usage.

File: applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h (hunk 12 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h:206]: As it is right now, the MVQN will only work with this one. Depending on how the design evolves (see my comment above) we could (ideally) have a unique `IBQN` class that works with several Jacobian approximators such as the IBQN-ILS (the point is that the interface equations are the same). Then we will need for sure to remove the `friend`.

File: applications/TrilinosApplication/custom_python/add_trilinos_convergence_accelerators_to_python.cpp (hunk 1 above)
Comment (rubenzorrilla) [reply applications/TrilinosApplication/custom_python/add_trilinos_convergence_accelerators_to_python.cpp]: Sure.

File: applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h (hunk 8 above)
Comment (philbucher) [reply applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h]: then I would take it out, I find it confusing if it doesn't actually change anything

**Review by philbucher — COMMENTED** [review]
minor rest is good to me 👍

File: applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h (hunk 8 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/custom_utilities/ibqn_mvqn_convergence_accelerator.h]: Done.

**Review by philbucher — DISMISSED** [review]
:+1:

File: applications/FSIApplication/tests/test_FSIApplication.py (hunk 5 above)
Comment (rubenzorrilla) [reply applications/FSIApplication/tests/test_FSIApplication.py]: Done.

**Review by philbucher — APPROVED** [review]

**Comment by rubenzorrilla** [comment]
I also included your minor comments about the testing (while doing so I realized that there was a test not being run :facepalm: ) and defaults of the IBQN-MVQN.

About the memory consumption... I've an ongoing implementation which should improve this a lot.
